pkgs/by-name/op/opencl-clang/llvm/llvm/default.nix

```diff
@@ -0,0 +1,369 @@
+      (fetchpatch {
+        url = "https://github.com/llvm/llvm-project/commit/2e1b838a889f9793d4bcd5dbfe10db9796b77143.patch";
+        relative = "llvm";
+        hash = "sha256-Ot45P/iwaR4hkcM3xtLwfryQNgHI6pv6ADjv98tgdZA=";
+      })
+
+      # Just like the `gnu-install-dirs` patch, but for `polly`.
+      ./gnu-install-dirs-polly.patch
+      # Just like the `llvm-lit-cfg` patch, but for `polly`.
+      ./polly-lit-cfg-add-libs-to-dylib-path.patch
+    ];
+
+    nativeBuildInputs = [
+      cmake
+      # while this is not an autotools build, it still includes a config.guess
+      # this is needed until scripts are updated to not use /usr/bin/uname on FreeBSD native
+      updateAutotoolsGnuConfigScriptsHook
+      python
+      ninja
+    ];
+
+    buildInputs = [
+      libxml2
+      libffi
+      libpfm
+    ];
+
+    propagatedBuildInputs = lib.optional (stdenv.buildPlatform == stdenv.hostPlatform) ncurses ++ [
+      zlib
+    ];
+
+    postPatch =
+      # FileSystem permissions tests fail with various special bits
+      ''
+        substituteInPlace unittests/Support/CMakeLists.txt \
+          --replace-fail "Path.cpp" ""
+        rm unittests/Support/Path.cpp
+        substituteInPlace unittests/IR/CMakeLists.txt \
+          --replace-fail "PassBuilderCallbacksTest.cpp" ""
+        rm unittests/IR/PassBuilderCallbacksTest.cpp
+        rm test/tools/llvm-objcopy/ELF/mirror-permissions-unix.test
+        rm utils/lit/tests/googletest-timeout.py
+        patchShebangs test/BugPoint/compile-custom.ll.py
+      '';
+
+    # Workaround for configure flags that need to have spaces
+    preConfigure = ''
+      cmakeFlagsArray+=(
+        -DLLVM_LIT_ARGS="--verbose -j''${NIX_BUILD_CORES}"
+      )
+    '';
+
+    # E.g. Mesa uses the build-id as a cache key (see #93946):
+    LDFLAGS = "-Wl,--build-id=sha1";
+
+    cmakeBuildType = "Release";
+
+    cmakeFlags =
+      let
+        # These flags influence llvm-config's BuildVariables.inc in addition to the
+        # general build. We need to make sure these are also passed via
+        # CROSS_TOOLCHAIN_FLAGS_NATIVE when cross-compiling or llvm-config-native
+        # will return different results from the cross llvm-config.
+        #
+        # Some flags don't need to be repassed because LLVM already does so (like
+        # CMAKE_BUILD_TYPE), others are irrelevant to the result.
+        flagsForLlvmConfig = [
+          (lib.cmakeFeature "LLVM_INSTALL_PACKAGE_DIR" "${placeholder "dev"}/lib/cmake/llvm")
+          (lib.cmakeBool "LLVM_ENABLE_RTTI" true)
+          (lib.cmakeBool "LLVM_LINK_LLVM_DYLIB" true)
+          (lib.cmakeFeature "LLVM_TABLEGEN" "${tblgen}/bin/llvm-tblgen")
+        ];
+      in
+      flagsForLlvmConfig
+      ++ [
+        (lib.cmakeBool "LLVM_INSTALL_UTILS" true) # Needed by rustc
+        (lib.cmakeBool "LLVM_BUILD_TESTS" finalAttrs.finalPackage.doCheck)
+        (lib.cmakeBool "LLVM_ENABLE_FFI" true)
```

Comment:
I’m sceptical that this and the `libffi` dependency are required for this use case? But I’m not sure.